### PR TITLE
DM-31226: Add functionality needed for WCSFitTask in drp_tasks

### PR DESCRIPTION
**Pull request by cmsaunders**
This PR adds the following functionality:

- Conversion from SIP to TPV for the input WCS
- More information about the pixel maps is exposed to the Python wrapper
- Options to output catalogs with residuals, fit star positions, etc., in the form of dictionaries that can be trivially converted to Pandas Dataframes.
- Improvements in interface for fitting proper motions and parallax.

### CODE REVIEW

File: astrometry/include/PixelMap.h [hunk 1]
```diff
@@ -256,6 +257,56 @@ namespace astrometry {
     virtual void write(YAML::Emitter& os) const;
 #endif
     virtual string getType() const {return type();}
+    Matrix33 getPixMatrix() {
+      cerr << "start get pix" << endl;
```
Comment (natelust) [inline astrometry/include/PixelMap.h]: Do we normally send things out to std err vs logging? Or is this just because our gbdes forking?

File: astrometry/include/PixelMap.h [hunk 2]
```diff
@@ -256,6 +257,56 @@ namespace astrometry {
     virtual void write(YAML::Emitter& os) const;
 #endif
     virtual string getType() const {return type();}
+    Matrix33 getPixMatrix() {
```
Comment (natelust) [inline astrometry/include/PixelMap.h]: Why are you defining all of this in the header, instead of the signature here and the body in a src file?

File: astrometry/include/PixelMap.h [hunk 3]
```diff
@@ -256,6 +257,56 @@ namespace astrometry {
...
+      cerr << "start get pix" << endl;
+      auto gnom = dynamic_cast<const astrometry::Gnomonic*>(pix);
+      cerr << "did dyn cast" << endl;
+      Matrix33 tmpMatrix;
+      tmpMatrix.setToIdentity();
```
Comment (natelust) [inline astrometry/include/PixelMap.h]: would it make sense to do this with tmpMatrix only if it is actually going to be needed? (same below)

File: astrometry/include/PixelMap.h [hunk 4]
```diff
@@ -256,6 +257,56 @@ namespace astrometry {
...
+      return gnom->getOrient()->m();
+    }
+    Vector2 getPixPole() {
+      auto gnom = dynamic_cast<const astrometry::Gnomonic*>(pix);
+      if (!gnom) {
+        auto icrs = dynamic_cast<const astrometry::SphericalICRS*>(pix);
+        if (!icrs) {
+          Vector2 lonlat;
+          cerr << "not gnom or icrs" << endl;
+          return lonlat;
```
Comment (natelust) [inline astrometry/include/PixelMap.h]: I'm not familiar with this object, is lonlat a class level variable?

File: astrometry/include/PixelMap.h [hunk 5]
```diff
@@ -256,6 +257,56 @@ namespace astrometry {
...
+        }
+        return icrs->getLonLat();
+      }
+      SphericalICRS pole = gnom->getOrient()->getPole();
+      Vector2 lonlat = pole.getLonLat();
```
Comment (natelust) [inline astrometry/include/PixelMap.h]: if lonlat is a class level variable name, maybe it would be a good idea to use a different variable name here for human readability

File: astrometry/include/PixelMapCollection.h [hunk 6]
```diff
@@ -233,6 +237,9 @@ namespace astrometry {
     // Includes self.  Assumes no dependence cycles.
     set<string> dependencies(string mapName) const;
 
+    // Produce a list giving the atomic transformation sequence needed to implement a
+    // specified map. Assumes no dependence cycles.
+    list<string> orderAtoms(string mapName) const;
```
Comment (natelust) [inline astrometry/include/PixelMapCollection.h:242]: why did you move this?

File: astrometry/src/PixelMapCollection.cpp [hunk 7]
```diff
@@ -135,6 +135,61 @@ PixelMapCollection::getParams() const {
...
+  if (mel.atom) {
+    mapType = mel.atom->getType();
+  }
+  else {
+    mapType = SubMap::type();
+  }
+  return mapType;
+}
+
+DVector PixelMapCollection::getWcsNativeCoords(string wcsName, bool degrees) const {
```
Comment (natelust) [inline astrometry/src/PixelMapCollection.cpp]: You seem to switch back and forth between the convention of preceding return type on it's own line, and before it on the same line, What is the convention in the package? If there is none, can you make your commits consistent

File: astrometry/src/PixelMapCollection.cpp [hunk 8]
```diff
@@ -135,6 +135,61 @@ PixelMapCollection::getParams() const {
...
+  auto wcspair = wcsElements.find(wcsName);
+  if (wcspair == wcsElements.end())
+    throw AstrometryError("Could not find wcsName " + wcsName);
+  
+  const WcsElement& wcs = wcspair->second;
+  const SphericalCustomBase *coords = dynamic_cast<const SphericalCustomBase*>(wcs.nativeCoords);
+  if (coords == nullptr) {
```
Comment (natelust) [inline astrometry/src/PixelMapCollection.cpp:181]: in previous changes you used !var to test for nullptr-ness, but here you use an explicit check. Do you have a reason for choosing either? If so I personally like the explicit nullptr check.

File: astrometry/src/PixelMapCollection.cpp [hunk 9]
```diff
@@ -135,6 +135,61 @@ PixelMapCollection::getParams() const {
...
+    throw AstrometryError("NativeCoords subclass does not have 'orient' property");
+  }
+  SphericalICRS pole = coords->getOrient()->getPole();
```
Comment (natelust) [inline astrometry/src/PixelMapCollection.cpp:184]: I see you are de-referencing the pointer returned by `getOrient()` is it possible for that to be null?

File: gbutil/src/Poly2d.cpp [hunk 10]
```diff
@@ -319,3 +319,23 @@ Poly2d::fillFromVector(const DVector& v) {
...
+  int iA, iB, jA, jB;
+  DVector AVector = getC();
+  DVector BVector = polyB.getC();
+  for (int m = 0; m < AVector.size(); m++) {
+    for (int n = 0; n < BVector.size(); n++) {
+      this->powersOfIndex(m, iA, jA);
+      polyB.powersOfIndex(n, iB, jB);
+      int outOrderX = iA + iB;
+      int outOrderY = jA + jB;
+      if ((outOrderX > orderX) || (outOrderY > orderY)) continue;
```
Comment (natelust) [inline gbutil/src/Poly2d.cpp]: What does DMatrix default initialized to? if you continue here, what happens to the entries that are not populated

File: include/Instrument.h [hunk 11]
```diff
@@ -81,15 +81,16 @@ class ExposuresHelper {
 public:
     ExposuresHelper(vector<string> expNames_, vector<int> fieldNumbers_, vector<int> instrumentNumbers_,
                     vector<double> ras_, vector<double> decs_, vector<double> airmasses_,
-                    vector<double> exposureTimes_, vector<double> mjds_)
+                    vector<double> exposureTimes_, vector<double> mjds_, vector<vector<double>> observatories_)
```
Comment (natelust) [inline include/Instrument.h:84]: Changing public API is a big thing, are you the only consumer of this? Have you considered having a default for backwards compatibility. Even if you are the only Rubin consumer, consider how this will affect your ability to upstream changes, and how it might lead to further code drift going forward.

File: include/Instrument.h [hunk 12]
```diff
@@ -113,6 +115,14 @@ class ExposuresHelper {
...
             expo->exptime = exposureTimes[i];
             expo->mjd = mjds[i];
+            if (fieldEpochs.size() > 0) {
+                // Also calculate time in years after field's reference epoch
+                astrometry::UT ut;
+                ut.setMJD(expo->mjd);
+                // Note that getTTyr returns years since J2000.
+                expo->pmTDB = ut.getTTyr() - (fieldEpochs[expo->field] - 2000.);
+            }
+            for (int j = 0; j < 3; ++j) expo->observatory[j] = observatories[i][j];
```
Comment (natelust) [inline include/Instrument.h:125]: I am not personally a fan of one line, or braceless for loops or if statements. I can't remember what our dev guide says though, can you double check on this?

File: pydir/wcsfit.cpp [hunk 13]
```diff
@@ -24,6 +24,30 @@ void declareBounds(py::module &m) {
     py::class_<Class>(m, "Bounds").def(py::init<const T, const T, const T, const T>());
 }
 
+template <typename T>
+py::array_t<T> vectorToNumpy(std::vector<T> const &vect) {
```
Comment (natelust) [inline pydir/wcsfit.cpp:28]: won't pybind do this for you? I thought they had some directive to do so, but maybe I am miss remembering, or it applies too broadly. Can you double check?

File: pydir/wcsfit.cpp [hunk 14]
```diff
@@ -52,7 +76,11 @@ PYBIND11_MODULE(wcsfit, m) {
                 astrometry::SphericalICRS pole(ra, dec);
                 astrometry::Orientation orientIn(pole);
                 return astrometry::Gnomonic(orientIn.getPole(), orientIn);
-            }));
+            }))
+            .def("getOrientMatrix", [](astrometry::Gnomonic &self) {
+                const astrometry::Orientation* orient=self.getOrient();
```
Comment (natelust) [inline pydir/wcsfit.cpp:81]: Can this be null?

File: pydir/wcsfit.cpp [hunk 15]
```diff
@@ -52,7 +76,11 @@ PYBIND11_MODULE(wcsfit, m) {
...
+                return orient->m();
```
Comment (natelust) [inline pydir/wcsfit.cpp:85]: can this be null?

File: src/subs/FitSubroutines.cpp [hunk 16]
```diff
@@ -1540,7 +1540,8 @@ void readObjects_oneExtension(const vector<unique_ptr<Exposure>> &exposures, int
     // Get fieldProjection for this catalog
     astrometry::SphericalCoords *fieldProjection = fieldProjections[expo.field]->duplicate();
     if (S::isAstro) {
-        errorColumnIsDouble = isDouble(ff, pmCatalog ? pmCovKey : xyErrKeys[0], -1);
+        if (pmCatalog) errorColumnIsDouble = isDouble(ff, pmCovKey, 0);
+        else errorColumnIsDouble = isDouble(ff, xyErrKeys[0], -1);
```
Comment (natelust) [inline src/subs/FitSubroutines.cpp:1544]: what if neither of these is double?

File: src/subs/FitSubroutines.cpp [hunk 17]
```diff
@@ -2234,17 +2235,17 @@ void Astro::saveResults(const astrometry::MCat &matches, string outCatalog, stri
             pmClip[i] = pmDetPtr->isClipped;
             pmReserve[i] = pmDetPtr->itsMatch->getReserved();
             vector<float> vv(5);
-            for (int i = 0; i < 5; i++) vv[i] = pmDetPtr->pmMean[i];
+            for (int ii = 0; ii < 5; ii++) vv[ii] = pmDetPtr->pmMean[ii];
             // Keep position in WCS_UNITS (degrees), put others into RESIDUAL_UNITS
-            vv[astrometry::VX] *= units[i];
-            vv[astrometry::VY] *= units[i];
-            vv[astrometry::PAR] *= units[i];
+            vv[astrometry::VX] *= units[astrometry::VX];
```
Comment (natelust) [inline src/subs/FitSubroutines.cpp:2240]: I cant tell if this is a github rendering, but is this inside the for loop?

File: src/subs/FitSubroutines.cpp [hunk 18]
```diff
@@ -2396,6 +2397,418 @@ void Astro::saveResults(const astrometry::MCat &matches, string outCatalog, stri
...
+            starParallax.push_back(pm[astrometry::PAR] * units[astrometry::PAR]);
+        }
+        else {
+            astrometry::Vector2 matchCentroid = m->predict();
+            astrometry::SphericalCoords *projection;
+            for (auto const &detptr : *m) {
+                projection = catalogProjections[detptr->catalogNumber];
+                if (projection) break;
+            }
+            if (projection) {
```
Comment (natelust) [inline src/subs/FitSubroutines.cpp]: couldn't you save this branch if you just did this work inside line 2772, and did break at the end?

File: src/subs/TPVMap.cpp [hunk 19]
```diff
@@ -244,6 +244,159 @@ unique_ptr<Wcs> astrometry::readTPV(const img::Header &h, string name) {
...
+    int a_order = header.find("A_ORDER")->second;
+    int b_order = header.find("B_ORDER")->second;
+
+    DMatrix ac(b_order+1, a_order+1, 0.0);
+    DMatrix bc(b_order+1, b_order+1, 0.0);
+
+    // Fill ac and bc
+    for (int m = 0; m <= a_order; m++) {
+        for (int n = 0; n <= (a_order - m); n++) {
+            std::string keyString = "A_" + std::to_string(m) + "_" + std::to_string(n);
```
Comment (natelust) [inline src/subs/TPVMap.cpp:309]: is addition the best way to do this? or is there a standard formatting?

File: src/subs/TPVMap.cpp [hunk 20]
```diff
@@ -244,6 +244,159 @@ unique_ptr<Wcs> astrometry::readTPV(const img::Header &h, string name) {
...
+            pv = new PolyMap(p1Identity, p2, polyName, Bounds<double>(-1., 1., -1., 1.), POLYSTEP);
+        }
+    }
+    list<PixelMap *> pmlist;
+    pmlist.push_back(&lm);
+    if (pv) pmlist.push_back(pv);
+    // Create a SubMap that owns a duplicate of these one or two maps (no sharing):
+    SubMap sm(pmlist, name, false);
+    // Delete the polymap if we made it:
+    delete pv;
```
Comment (natelust) [inline src/subs/TPVMap.cpp:393]: I'm not sure I follow all the need for allocating a new object, and freeing like this. Couldn't this be refactored to not need the dynamic memory like this? perhaps by directly creating the SubMap?

File: src/subs/WCSFitRoutine.cpp [hunk 21]
```diff
@@ -352,11 +365,15 @@ void WCSFit::setObjects(int i, const map<std::string, std::vector<double>> &tabl
                         const std::string &idKey, const std::string &pmCovKey, const std::string &magKey,
                         const int &magKeyElement, const std::string &magErrKey, const int &magErrKeyElement,
                         const std::string &pmRaKey, const std::string &pmDecKey,
-                        const std::string &parallaxKey) {
+                        const std::string &parallaxKey, const std::vector<std::vector<double>> &pmCov) {
```
Comment (natelust) [inline src/subs/WCSFitRoutine.cpp:368]: Does this parameter have a default in a header somewhere?

**Review by natelust — APPROVED** [review]
There is some work to be done here, and I would be happy to weight in on any responses, or look again if you wish, but I am also happy for you to take these as advisement and fix it up how you decide.

File: astrometry/include/PixelMap.h (hunk 2 above)
Comment (cmsaunders) [reply astrometry/include/PixelMap.h]: I probably did this before I read that section in the dev guide...will fix.

File: astrometry/include/PixelMap.h (hunk 1 above)
Comment (cmsaunders) [reply astrometry/include/PixelMap.h]: There's a separate ticket to move to proper logging. However, in this case this should just be removed.

File: astrometry/include/PixelMap.h (hunk 4 above)
Comment (cmsaunders) [reply astrometry/include/PixelMap.h]: I just removed this function, because it was something I used for debugging, and I don't think I will need it again.

File: astrometry/include/PixelMap.h (hunk 5 above)
Comment (cmsaunders) [reply astrometry/include/PixelMap.h]: See above.

File: astrometry/include/PixelMapCollection.h (hunk 6 above)
Comment (cmsaunders) [reply astrometry/include/PixelMapCollection.h:242]: Not super necessary, but it's very related to `dependencies`, so I thought it made more sense to have them together. I was also considering adding more related functions.

File: astrometry/src/PixelMapCollection.cpp (hunk 7 above)
Comment (cmsaunders) [reply astrometry/src/PixelMapCollection.cpp]: Yeah, the package is not consistent, though I think within a single file there is consistency. I'll try to make my commits match the file.

File: astrometry/src/PixelMapCollection.cpp (hunk 8 above)
Comment (cmsaunders) [reply astrometry/src/PixelMapCollection.cpp:181]: In some places I'm following existing code, but where that's not the case I'll switch to the explicit nullptr check.

File: astrometry/src/PixelMapCollection.cpp (hunk 9 above)
Comment (cmsaunders) [reply astrometry/src/PixelMapCollection.cpp:184]: I don't think so--it's set in the constructor in line 179 right above, and I don't think the value that goes into it could be null.

File: gbutil/src/Poly2d.cpp (hunk 10 above)
Comment (cmsaunders) [reply gbutil/src/Poly2d.cpp]: This should actually throw an error, because `outOrderX` can't be greater than `orderX`.

File: include/Instrument.h (hunk 11 above)
Comment (cmsaunders) [reply include/Instrument.h:84]: Yeah, this is a class I added specifically for pybind11 so that the `Exposure` objects aren't exposed to python. No one else is using it yet.

File: include/Instrument.h (hunk 12 above)
Comment (cmsaunders) [reply include/Instrument.h:125]: Dev guide says it's okay if it's all on one line: https://developer.lsst.io/cpp/style.html#single-statement-if-else-for-or-while-statements-must-only-be-written-without-brackets-if-on-one-line

File: pydir/wcsfit.cpp (hunk 13 above)
Comment (cmsaunders) [reply pydir/wcsfit.cpp:28]: This is what Jim and I worked out to return a dictionary with mixed-type values. Before doing this, it was converting `std::map<std::string, std::vector<double>>` to a dictionary of lists, and didn't work with all the items I needed to include.

File: pydir/wcsfit.cpp (hunk 14 above)
Comment (cmsaunders) [reply pydir/wcsfit.cpp:81]: I don't think so, but do you think it would be safer to check here? For this and the next comment, they are set in the constructor and are private. Do you think there is still the danger that they could become null?

File: pydir/wcsfit.cpp (hunk 15 above)
Comment (cmsaunders) [reply pydir/wcsfit.cpp:85]: See above.

File: src/subs/FitSubroutines.cpp (hunk 16 above)
Comment (cmsaunders) [reply src/subs/FitSubroutines.cpp:1544]: This is checking whether the error is double, so it's okay if it's not. It just uses checks a different error parameter if the catalog has proper motions.

File: src/subs/FitSubroutines.cpp (hunk 17 above)
Comment (cmsaunders) [reply src/subs/FitSubroutines.cpp:2240]: Yes. I changed this because I saw that `i` was being reused as an index. I guess it wasn't affecting the `i` in the outer scope, because I think the vector was still being filled, but it seems like bad practice.

File: src/subs/TPVMap.cpp (hunk 19 above)
Comment (cmsaunders) [reply src/subs/TPVMap.cpp:309]: This seems like the simplest way in c++17. I spent some time looking and didn't see better options, but please tell me if I'm missing something.

File: src/subs/TPVMap.cpp (hunk 20 above)
Comment (cmsaunders) [reply src/subs/TPVMap.cpp:393]: I basically copied what is in `readTPV`. We can discuss how to refactor.

File: src/subs/WCSFitRoutine.cpp (hunk 21 above)
Comment (cmsaunders) [reply src/subs/WCSFitRoutine.cpp:368]: Yeah, it defaults to zero. Why do you ask, because it changes the API?
